File: module-services/src/main/java/eu/geekhome/services/hardware/HardwareAdapter.java
package eu.geekhome.services.hardware;

import com.geekhome.common.*;
import eu.geekhome.services.events.EventsSink;

import java.util.Calendar;
import java.util.List;

public interface HardwareAdapter {
    String getId();
    void discover(PortIdBuilder idBuilder, List<Port> ports, EventsSink<String> eventsSink);
    boolean canBeRediscovered();
    void refresh(Calendar now) throws Exception;
    AdapterState getState();
    void resetLatches();
    void reconfigure(OperationMode operationMode) throws Exception;
    long getLastDiscoveryTime();
    Throwable getLastError();
}

File: module-services/src/main/java/eu/geekhome/services/hardware/HardwareAdapter.kt
package eu.geekhome.services.hardware

import eu.geekhome.services.events.EventsSink
import java.util.*

interface HardwareAdapter {

    val newPorts: List<Port<*>>
    fun clearNewPorts()
    val id: String
    suspend fun discover(discoverySink: EventsSink) : MutableList<Port<*>>

    @Throws(Exception::class)
    fun refresh(now: Calendar)
    var state: AdapterState
    fun executePendingChanges()

    @Throws(Exception::class)
    fun reconfigure(operationMode: OperationMode)
    fun stop()
    fun start(operationSink: EventsSink)

    var lastDiscoveryTime: Long
    var lastError: Throwable?
}